File: 250430/방향에 맞춰 최대로 움직이기/max-movements-with-direction.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int n;
int num[5][5];
int d[5][5];
int r, c;
int dir[9][2] = { {0, 0}, {-1, 0}, {-1, 1}, {0, 1}, {1, 1},
                {1, 0}, {1, -1}, {0, -1}, {-1, -1}}; 

int result = 0;
void func(int x, int y, int cnt) {

    bool find = 0;
    for (int i = 1; i <= n; i++) {
        int nx = x + dir[d[x][y]][0] * i;
        int ny = y + dir[d[x][y]][1] * i;
        if(nx < 1 || ny < 1 || nx > n || ny > n) continue;
        if (num[x][y] < num[nx][ny]) {
            find = 1;
            func(nx, ny, cnt + 1);
        }
    }

    if (!find) {
        result = max(result, cnt);
        return;
    }
}
int main() {
    cin >> n;

    for (int i = 1; i <= n; i++)
        for (int j = 1; j <= n; j++) cin >> num[i][j];

    for (int i = 1; i <= n; i++)
        for (int j = 1; j <= n; j++) cin >> d[i][j];


    cin >> r >> c;
    func(r,c,0);
    cout << result;

    // Please write your code here.

    return 0;
}

File: 250807/방향에 맞춰 최대로 움직이기/max-movements-with-direction.py
n = int(input())
num = [list(map(int, input().split())) for _ in range(n)]
move_dir = [list(map(int, input().split())) for _ in range(n)]
r, c = map(int, input().split())
r -= 1
c -= 1
ans = 0

# Please write your code here.
dirs = [(-1, 0), (-1, 1), (0, 1), (1, 1), (1, 0), 
        (1, -1), (0, -1), (-1, -1)]

def in_range(x, y):
    return 0 <= x < n and 0 <= y < n

def backtracking(r, c, move_cnt):
    global ans
    ans = max(ans, move_cnt)

    dx, dy = dirs[move_dir[r][c] - 1]

    for dist in range (1, n):
        nx = r + dx * dist
        ny = c + dy * dist
        # print(f"next is {nx} and {ny}")
        if not in_range(nx, ny) or num[r][c] > num[nx][ny]:
            continue

        backtracking(nx, ny, move_cnt + 1)
    return

backtracking(r, c, 0)
print(ans)